R-package/src/xgboost_R.cc

```diff
@@ -540,15 +546,147 @@ XGB_DLL SEXP XGDMatrixNumCol_R(SEXP handle) {
   return ScalarInteger(static_cast<int>(ncol));
 }
 
+XGB_DLL SEXP XGDuplicate_R(SEXP obj) {
+  return Rf_duplicate(obj);
+}
+
+XGB_DLL SEXP XGDuplicateAttrib(SEXP src, SEXP dest) {
+  DUPLICATE_ATTRIB(dest, src);
+  return R_NilValue;
+}
+
+XGB_DLL SEXP XGPointerEqComparison(SEXP obj1, SEXP obj2) {
+  return Rf_ScalarLogical(R_ExternalPtrAddr(obj1) == R_ExternalPtrAddr(obj2));
+}
+
 // functions related to booster
-void _BoosterFinalizer(SEXP ext) {
-  if (R_ExternalPtrAddr(ext) == NULL) return;
-  CHECK_CALL(XGBoosterFree(R_ExternalPtrAddr(ext)));
-  R_ClearExternalPtr(ext);
+static void _BoosterFinalizer(SEXP R_ptr) {
```

Comment:
Instead of using `static`, can we put all the internal functions in an anonymous namespace?

Reply:
Moved to a namespace.